transport/http_transport.go

```diff
@@ -0,0 +1,467 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package transport
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/groupcache/groupcache-go/v3/transport/pb"
+	"github.com/groupcache/groupcache-go/v3/transport/peer"
+	"golang.org/x/net/proxy"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	DefaultBasePath = "/_groupcache/"
+	defaultScheme   = "http"
+)
+
+var bufferPool = sync.Pool{
+	New: func() interface{} { return new(bytes.Buffer) },
+}
+
+type GroupCacheInstance interface {
+	GetGroup(string) Group
+}
+
+type Logger interface {
+	Info(msg string, args ...any)
+	Error(msg string, args ...any)
+}
+
+type Transport interface {
+	// New returns a clone of this instance suitable for passing to groupcache.New()
+	// Example usage:
+	//
+	// transport := groupcache.NewHttpTransport(groupcache.HttpTransportOptions{})
+	// groupcache.New(groupcache.Config{Transport: transport.New()})
+	New() Transport
+
+	// Register registers the provided *Instance with the HttpTransport.
+	//
+	// This method sets the instance field of the HttpTransport to the provided instance.
+	// The instance is used by the ServeHTTP method to serve groupcache requests.
+	Register(instance GroupCacheInstance)
+
+	// NewClient returns a new Client suitable for the transport implementation. The client returned is used to communicate
+	// with a specific peer. This method will be called for each peer in the peer list when groupcache.Instance.SetPeers() is
+	// called.
+	NewClient(ctx context.Context, peer peer.Info) (peer.Client, error)
+
+	// SpawnServer spawns a server that will handle incoming requests for this transport
+	// This is used by daemon and cluster packages to create a cluster of instances using
+	// this specific transport.
+	SpawnServer(ctx context.Context, address string) error
+
+	// ShutdownServer shuts down the server started when calling SpawnServer()
+	ShutdownServer(ctx context.Context) error
+
+	// ListenAddress returns the address the server is listening on after calling SpawnServer().
+	ListenAddress() string
+}
+
+// HttpTransportOptions options for creating a new HttpTransport
+type HttpTransportOptions struct {
+	// Context (Optional) specifies a context for the server to use when it
+	// receives a request.
+	// defaults to http.Request.Context()
+	Context func(*http.Request) context.Context
+
+	// Client (Optional) provide a custom http client with TLS config.
+	// defaults to http.DefaultClient
+	Client *http.Client
+
+	// Scheme (Optional) is is either `http` or `https`
+	// defaults to `http`
+	Scheme string
+
+	// BasePath (Optional) specifies the HTTP path that will serve groupcache requests.
+	// defaults to "/_groupcache/".
+	BasePath string
+
+	// Logger
+	Logger Logger
+
+	// TODO: Support for TLS
```

Comment:
May want to mark on the type itself that TLS is a TODO as Client and Scheme imply that it is supported where the TODO is internal to the code 👀 Otherwise code completion misses this note